src/raygun.rum.js

```diff
@@ -119,79 +167,139 @@ var raygunRumFactory = function (window, $, Raygun) {
               }
         };
 
-        this.setUser = function (user) {
-            self.user = user;
+        this.sendPerformance = function (firstLoad) {
+            var performanceData = getPerformanceData(this.virtualPage, firstLoad);
+
+            if (performanceData === null || performanceData.length < 0) {
+                return;
+            }
+
+            addPerformanceTimingsToQueue(performanceData, false);
         };
 
-        this.withTags = function (tags) {
-            self.tags = tags;
+        this.sendChildAssets = function(forceSend) {
+          if (forceSend) {
+            processVirtualPageTimingsInQueue();
+          }
+          var data = [];
+          extractChildData(data);
+          addPerformanceTimingsToQueue(data, forceSend);
         };
 
-        this.endSession = function () {
+        this.sendQueuedItems = function() {
+            if (self.queuedItems.length > 0) {
+                // Dequeue:
+                self.queuedItems = sortCollectionByProperty(self.queuedItems, 'timestamp');
+                var itemsToSend = self.queuedItems.splice(0, self.maxQueueItemsSent);
+
+                self.sendItemsImmediately(itemsToSend);
+            }
+        };
+
+        this.sendItemImmediately = function(item) {
+            var itemsToSend = [item];
+            self.sendItemsImmediately(itemsToSend);
+        };
+
+        this.sendItemsImmediately = function(itemsToSend) {
             var payload = {
-                eventData: [{
-                  sessionId: self.sessionId,
-                  timestamp: new Date().toISOString(),
-                  type: 'session_end'
-                }]
+                eventData: itemsToSend
             };
 
-            self.postPayload(payload);
-        };
+            var successCallback = function() {
+                log('Raygun4JS: Items sent successfully. Queue length: ' + self.queuedItems.length);
+            };
 
-        this.heartBeat = function () {
+            var errorCallback = function(response) {
+                // Requeue:
+                self.requeueItemsToFront(itemsToSend);
```

Comment:
Won't this just result in failed requests (that would keep failing) being sent forever?

Reply:
No this shouldn't happen, the event items get dequeued before sending and only requeued if the request fails. So once the request succeeds the items will no longer be in the queue.

Reply:
Yes, but if it fails and gets put onto the queue and then fails again (say because the app is rate limited, or some network condition) it will get put back onto the queue again, forever?